hail/python/hail/linalg/blanczos_data_script.py

```diff
@@ -0,0 +1,265 @@
+    mt = mt.select_entries(x = field)
+    ht = mt.localize_entries(entries_array_field_name='entries')
+    # now ht.entries is an array of structs with one field, x
+
+    # we'll also want to mean-impute/variance-normalize/etc here
+    ht = ht.select(xs = ht.entries.map(lambda e: e['x']))
+    # now ht.xs is an array of float64
+
+    # now need to produce groups of G
+    ht = ht.add_index()
+    ht = ht.group_by(row_group_number= hl.int32(ht.idx // group_size)) \
+        .aggregate(ndarray=hl.nd.array(hl.agg.collect(ht.xs)))
+    # may require a .T on ndarray
+
+    return ht.checkpoint(tmp_path, overwrite=True)
+
+def chunk_ndarray(a, group_size):
+    n_groups = ceil(a.shape[0] / group_size)
+    groups = []
+    for i in range(n_groups):
+        start = i * group_size
+        if i == n_groups - 1:
+            end = a.shape[0]
+        else:
+            end = (i + 1) * group_size
+        groups.append(a[start:end, :])
+    return groups
+
+
+# Concatenate the ndarrays with a blocked Table
+def concatBlocked(A):
+    blocks = A.ndarray.collect()
+    big_mat = np.concatenate(blocks, axis=0)
+    ht = ndarray_to_table([big_mat])
+    return ht
+
+def concatToNumpy(A):
+    blocks = A.ndarray.collect()
+    big_mat = np.concatenate(blocks, axis=0)
+    return big_mat
+
+def ndarray_to_table(chunked_arr):
+    structs = [hl.struct(row_group_number = idx, ndarray = block)
+               for idx, block in enumerate(chunked_arr)]
+    ht = hl.Table.parallelize(structs)
+    ht = ht.key_by('row_group_number')
+    return ht
+
+# function to multiply two blocks, given the two blocks
+# returns struct in form of array but not ndarray, includes the shape in the struct
+# to change the result product directly back into a ndarray we need to use from_column_major
+def block_product(left, right):
+    product = left @ right     
+    n_rows, n_cols = product.shape
+    return hl.struct(
+        shape=product.shape,
+        block=hl.range(hl.int(n_rows * n_cols)).map(
+            lambda absolute: product[absolute % n_rows, absolute // n_rows]))
+
+# takes in output of block_product
+def block_aggregate(prod):
+    shape = prod.shape
+    block = prod.block
+    return hl.nd.from_column_major(
+        hl.agg.array_sum(block),
+        hl.agg.take(shape, 1)[0])
+
+# returns flat array
+def to_column_major(ndarray):
+    n_rows, n_cols = ndarray.shape
+    return hl.range(hl.int(n_rows * n_cols)).map(
+        lambda absolute: ndarray[absolute % n_rows, absolute // n_rows])
+
+
+
+def matmul_rowblocked_nonblocked(A, B):
+    temp = A.annotate_globals(mat = B)
+    temp = temp.annotate(ndarray = temp.ndarray @ temp.mat)
+    temp = temp.select(temp.ndarray)
+    temp = temp.drop(temp.mat)
+    return temp
+
+def matmul_colblocked_rowblocked(A, B):
+    temp = A.transmute(ndarray = block_product(A.ndarray.transpose(), B[A.row_group_number].ndarray))
+    result_arr_sum = temp.aggregate(block_aggregate(temp.ndarray))
```

Comment:
This code can now be rewritten as:

```
def matmul_colblocked_rowblocked(A, B):
    temp = A.transmute(ndarray = A.ndarray.transpose() @ B[A.row_group_number].ndarray)
    result_arr_sum = temp.aggregate(hl.agg.ndarray_sum(temp.ndarray)
    return result_arr_sum
```

which should allow you to get rid of `block_aggregate`, `block_product`, and `to_column_major` altogether.